src/istio/dynamic_encoding/field_encoder.h

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ISTIO_DYNAMIC_ENCODING_FIELD_ENCODER_H
+#define ISTIO_DYNAMIC_ENCODING_FIELD_ENCODER_H
+
+#include <memory>
+#include <string>
+
+#include "encoder.h"
+
+#include "google/protobuf/descriptor.h"
+#include "google/protobuf/stubs/statusor.h"
+
+namespace istio {
+namespace dynamic_encoding {
+
+class FieldEncoder : public Encoder {
+ private:
+  // proto key  -- EncodeVarInt ((field_number << 3) | wire_type)
+  std::string protoKey;
+  // if packed this is set to true
+  bool packed;
+
+  // packed fields have a list of encoders.
+  // non-packed fields have one.
+  std::set<std::unique_ptr<Encoder>> encoders;
+
+  // number fields are sorted by fieldEncoder number.
+  int number;
+  // name for debug.
+  std::string name;
+
+  // dynamic expr to be evaluated at runtime.
+  std::string expr;
+  const ::google::protobuf::FieldDescriptor* fieldDescriptor;
+
+  const ::google::protobuf::EnumDescriptor* enumDescriptor;
+
+
+ public:
+  FieldEncoder() {}
+  ~FieldEncoder() {}
+  // This takes an attribute bag and evaluates the expression.
+  ::google::protobuf::util::StatusOr<std::string> Encode(
+       std::map<std::string, std::string> bag) {
```

Comment:
Is it assumed that the value of the attribute is always a string?

Reply:
Didn't get to this part of using attributes yet :( Can change to absl::any ?

Reply:
That's good with me. This value is coming from expression language runtime, so I'd like to keep this open-ended for now. I think it could be proto value.

Reply:
Thanks! Fixed

Reply:
Can you make the whole bag type parametric? I'm not sure whether a map to any is appropriate in C++. I think the parametricity should not affect the encoding bits.

Reply:
If we make a map parametric, won't it mean that all the values should be of the same type... absl::any will give us freedom to have different value types in the map...

I am using a map of absl::any when getting data too...

Reply:
I was thinking of the bag type. E.g. Encoder<T> that takes something of type T and produces bytes.

Reply:
so something like:

template <typename T> 
::google::protobuf::util::StatusOr<char []> Encode(
       T bag)

Reply:
**

template <typename T> 
::google::protobuf::util::StatusOr<char []> Encode(
T bag)

Reply:
Yeap

Reply:
@kyessenov : Message in C++ has a built-in function ParseFromString :https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.message_lite#MessageLite.ParseFromString.details
Thus I feel it would be better for this function to return std::string...

Reply:
Agree about the output being a string. I disagree that the input should be a string. That would require serialization to call a function, which seems unnecessary. With a templatized version, we can use encoders from internal protobuf representation, and defer choosing an attribute bag representation.